### PR TITLE
Fix Nx.Defn.Expr.as_type to properly cast constants

### PR DESCRIPTION
**Pull request by jonatanklosko**
See the corresponding test.

### CODE REVIEW

File: nx/lib/nx/defn/expr.ex [hunk 1]
```diff
@@ -984,9 +984,16 @@ defmodule Nx.Defn.Expr do
...
     number =
-      if is_integer(number) and Nx.Type.float?(type),
-        do: Complex.multiply(1.0, number),
-        else: number
+      cond do
+        is_integer(number) and Nx.Type.float?(type) ->
+          Complex.multiply(1.0, number)
+
+        is_float(number) and Nx.Type.integer?(type) ->
+          floor(number)
```
Comment (josevalim) [inline nx/lib/nx/defn/expr.ex:992]: I don't think we should do this here, it most often is a bug. I think here we should raise and we do explicit conversions on as_type.

Comment (jonatanklosko) [reply nx/lib/nx/defn/expr.ex:992]: I think it's a valid case, see

https://github.com/elixir-nx/nx/blob/1f4ddfea59849da1291083cac51177915511d0f5/nx/lib/nx/defn/expr.ex#L586-L587

This is an optimisation when casting constants, so we need to cast the actual value accordingly.

Comment (josevalim) [reply nx/lib/nx/defn/expr.ex:992]: Yes, the bug is in `as_type`. We can cast from integer -> float, but never from float to integer transparently, unless using `as_type`. So the logic belongs there! :)

File: nx/lib/nx/defn/expr.ex [hunk 2]
```diff
@@ -984,9 +991,16 @@ defmodule Nx.Defn.Expr do
...
     number =
-      if is_integer(number) and Nx.Type.float?(type),
-        do: Complex.multiply(1.0, number),
-        else: number
+      cond do
+        is_integer(number) and Nx.Type.float?(type) ->
+          Complex.multiply(1.0, number)
+
+        not is_integer(number) and Nx.Type.integer?(type) ->
+          raise "value #{inspect(number)} is not valid for constant of type #{inspect(type)}"
```
Comment (jonatanklosko) [inline nx/lib/nx/defn/expr.ex:999]: Let me know if having this assertion is worth it.

File: nx/lib/nx/defn/expr.ex [hunk 3]
```diff
@@ -584,6 +584,13 @@ defmodule Nx.Defn.Expr do
     tensor = to_expr(tensor)
 
     if c = maybe_constant(tensor) do
+      c =
+        if is_float(c) and Nx.Type.integer?(out.type) do
+          floor(c)
```
Comment (josevalim) [inline nx/lib/nx/defn/expr.ex:589]: what the binary backend uses is:

```suggestion
          trunc(c)
```

File: nx/lib/nx/defn/expr.ex (hunk 2 above)
Comment (josevalim) [reply nx/lib/nx/defn/expr.ex:999]:
```suggestion
        not is_integer(number) and Nx.Type.integer?(type) ->
          raise ArgumentError, "value #{inspect(number)} is not valid for constant of type #{inspect(type)}"
```

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**